### PR TITLE
Allowing people to specify cron tasks for k8s-service

### PR DESCRIPTION
**Pull request by juandiegopalomino**
# Description
Turned out to be pretty straightforward. Details are in the markdown.

# Safety checklist
* [x] This change is backwards compatible and safe to apply by existing users
* [x] This change will NOT lead to data loss
* [x] This change will NOT lead to downtime who already has an env/service setup

## How has this change been tested, beside unit tests?
manually

### CODE REVIEW

File: modules/opta-k8s-service-helm/templates/crons.yaml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+          tolerations:
+          {{ range $toleration :=  $.Values.tolerations }}
+          - key: {{ $toleration.key | quote }}
+            value: {{ default "opta" $toleration.value | quote }}
+            effect: {{ default "NoSchedule" $toleration.effect | quote }}
+          {{ end }}
+          {{- end }}
+          containers:
+            - name: cron
+              image: "{{ $.Values.image }}"
```
Comment (bigbitbus) [inline modules/opta-k8s-service-helm/templates/crons.yaml:28]: So if we have a "cron" in the opta service file in a k8s-service module, then we are installing a "deployment" (the current behavior) and this "cronjob" resource into Kubernetes with the identical image and environment in both the deployment and the cronjob. That should be useful in e.g. running migrations or backups etc. periodically. 

But...then we have wedded the two aspects; so when one changes, the other does too. Should be fine but its a little unconventional to combine them. Although I cant think of another way given that the k8s-service deploy command accepts a single image name/tag only so for example we dont have the possibility of creating two different k8s-service modules, one for deployment and one for cronjob.  

I'd rather use the k8s-resource and define a separate cronjob there, but thats just a choice.

File: modules/opta-k8s-service-helm/templates/crons.yaml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+              imagePullPolicy: Always
+              command:
+              {{ range $command := $cron.commands }}
+              - {{ $command | quote }}
+              {{ end }}
+              env:
+              - name: RDS_CA_PATH # This is the path to the public key for the docdb tls
```
Comment (bigbitbus) [inline modules/opta-k8s-service-helm/templates/crons.yaml:35]: Why is this specific RDS_CA_PATH a special case to be defined inside the template; separate from the other envVars?

**Review by bigbitbus — COMMENTED** [review]
Looks like a practical solution; some comments to think of included.

File: modules/opta-k8s-service-helm/templates/crons.yaml (hunk 2 above)
Comment (juandiegopalomino) [reply modules/opta-k8s-service-helm/templates/crons.yaml:35]: needed for documentdb from way back in the day. Super annoying and we might want to consider removing it lol

File: modules/opta-k8s-service-helm/templates/crons.yaml (hunk 1 above)
Comment (juandiegopalomino) [reply modules/opta-k8s-service-helm/templates/crons.yaml:28]: Great insight, 2 points:
1. So we can totally just have a pure cronjob service by specifying the min/max containers to be 0-- then only the cronjobs would run!
2. The special treatment we've given to the image has bitten us so much in the past, but it looks from today's discussion and plans for the onsite that we'll be trying to purge it very soon

File: modules/aws_k8s_service/aws-k8s-service.md [hunk 3]
```diff
@@ -115,3 +115,32 @@ said nodes unless they have a matching toleration. Simply provide a list of desi
...
+    port:
+      http: 80
+    crons:
+      - commands:
+        - /bin/sh
+        - -c
+        - 'echo "Hello world!"'
+        schedule: "* * * * *"
+        env_vars:
+          CRON: "1"
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply modules/aws_k8s_service/aws-k8s-service.md:145]: if for whatever reason they want to pass extra env vars.

[the remark this answers is not in the document]
Comment (bigbitbus) [reply modules/aws_k8s_service/aws-k8s-service.md:145]: ok.

File: modules/opta-k8s-service-helm/templates/crons.yaml (hunk 1 above)
Comment (bigbitbus) [reply modules/opta-k8s-service-helm/templates/crons.yaml:28]: Ok

**Review by bigbitbus — APPROVED** [review]
approved after comments discussed.

File: config/registry/schemas/common-types/cron.json [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+      },
+      "description": "The command array to execute",
+      "pattern": "^[a-zA-Z0-9-]+$"
+    },
+    "schedule": {
+      "type": "string",
+      "description": "The crontab schedule for this job",
+      "pattern": "^[a-zA-Z0-9 /*-]+$"
+    },
+    "env_vars": {
```
Comment (RemyDeWolf) [inline config/registry/schemas/common-types/cron.json:18]: I'm questioning if we need this?
why not reuse the same env_vars that is set on the main container?
would that make it so the user doesn't need to update the env vars at 2 places?
or do you have a good use case where we need 2 different set of env vars?

Comment (juandiegopalomino) [reply config/registry/schemas/common-types/cron.json:18]: I was thinking sometimes a user would like to specify that this is a cron job, but I'm happy to remove it if you think it's a bad idea

File: config/registry/schemas/common-types/cron.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+      "items": {
+        "type": "string"
+      },
+      "description": "The command array to execute",
+      "pattern": "^[a-zA-Z0-9-]+$"
+    },
+    "schedule": {
+      "type": "string",
+      "description": "The crontab schedule for this job",
+      "pattern": "^[a-zA-Z0-9 /*-]+$"
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply config/registry/schemas/common-types/cron.json]: done

File: modules/aws_k8s_service/aws-k8s-service.md [hunk 6]
```diff
@@ -115,3 +115,32 @@ said nodes unless they have a matching toleration. Simply provide a list of desi
 
 Please refer to the taints specified in your environment Opta manifests to know what matching tolerations are right 
 for you.
+
+### Crons
+
+Opta gives you the option of adding a list of cron jobs to run as part of this service. This is done via the `crons`
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply modules/aws_k8s_service/aws-k8s-service.md]: done

File: config/registry/schemas/common-types/cron.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+  "$id": "https://app.runx.dev/common-types/cron",
+  "type": "object",
+  "properties": {
+    "commands": {
+      "type": "array",
+      "items": {
+        "type": "string"
+      },
+      "description": "The command array to execute",
+      "pattern": "^[a-zA-Z0-9-]+$"
```
[the remark this answers is not in the document]
Comment (juandiegopalomino) [reply config/registry/schemas/common-types/cron.json]: done

File: modules/opta-k8s-service-helm/templates/crons.yaml (hunk 1 above)
Comment (juandiegopalomino) [reply modules/opta-k8s-service-helm/templates/crons.yaml:28]: done

File: config/registry/schemas/common-types/cron.json (hunk 4 above)
Comment (RemyDeWolf) [reply config/registry/schemas/common-types/cron.json:16]: my preference would be to remove it (and from doc) to keep it as simple as possible. if a user comes with this requirement it's easy to add 👍

File: modules/gcp_k8s_service/gcp-k8s-service.md [hunk 8]
```diff
@@ -116,3 +116,36 @@ said nodes unless they have a matching toleration. Simply provide a list of desi
 Please refer to the taints specified in your environment Opta manifests to know what matching tolerations are right
 for you.
 
+### Crons
```
Comment (RemyDeWolf) [inline modules/gcp_k8s_service/gcp-k8s-service.md:119]: rename this title too

**Review by RemyDeWolf — APPROVED** [review]
2 minor follow up comments but overall looks great

File: config/registry/schemas/common-types/cron.json [hunk 9]
```diff
@@ -0,0 +1,21 @@
+{
+  "$id": "https://app.runx.dev/common-types/cron",
```
Comment (RemyDeWolf) [inline config/registry/schemas/common-types/cron.json:2]: rename the file and here too?

File: config/registry/schemas/common-types/cron.json (hunk 4 above)
Comment (RemyDeWolf) [reply config/registry/schemas/common-types/cron.json:16]: did you mean to remove it?

File: modules/aws_k8s_service/aws-k8s-service.yaml [hunk 10]
```diff
@@ -190,7 +190,15 @@ inputs:
     validator: list(include('toleration'), required=False)
     description: Taint tolerations to add to the pods.
     default: []
+  - name: cron_jobs
+    user_facing: true
+    validator: list(include('cron'), required=False)
+    description: A list of cronjobs to execute as part of this service
+    default: []
 extra_validators:
+  cron:
```
Comment (RemyDeWolf) [inline modules/aws_k8s_service/aws-k8s-service.yaml:199]: rename here also?

File: modules/gcp_k8s_service/gcp-k8s-service.yaml [hunk 11]
```diff
@@ -181,7 +181,15 @@ inputs:
     validator: list(include('toleration'), required=False)
     description: Taint tolerations to add to the pods.
     default: []
+  - name: cron_jobs
+    user_facing: true
+    validator: list(include('cron'), required=False)
+    description: A list of cronjobs to execute as part of this service
+    default: []
 extra_validators:
+  cron:
```
Comment (RemyDeWolf) [inline modules/gcp_k8s_service/gcp-k8s-service.yaml:190]: rename here?

**Pull request merged**